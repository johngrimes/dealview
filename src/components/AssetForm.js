import React from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'
import _ from 'lodash'

import InputField from './InputField.js'
import CurrencyField from './CurrencyField.js'
import HiddenField from './HiddenField.js'
import DateField from './DateField.js'
import * as Validations from '../utils/formValidation.js'

import '../styles/forms.css'
import '../styles/buttons.css'
import '../styles/tables.css'

class AssetForm extends React.Component {
  static propTypes = {
    asset: PropTypes.shape({
      name: PropTypes.string,
      currentValue: PropTypes.string,
      purchaseDate: PropTypes.string,
      purchaseAmount: PropTypes.string,
      saleDate: PropTypes.string,
      forecastGrowth: PropTypes.string,
    }),
    onSubmit: PropTypes.func,
  }

  constructor(props) {
    super(props)
    this.state = {
      asset: {},
      current: true,
      allErrorsShown: false,
      focusedInput: 'name',
    }

    if (props.asset) {
      const asset = props.asset
      this.state = {
        ...this.state,
        asset,
        current:
          typeof asset.purchaseDate !== 'string' &&
          typeof asset.purchaseAmount !== 'number',
        errors: AssetForm.validate(asset),
        allErrorsShown: true,
      }
    }

    this.handleFocus = this.handleFocus.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleChange(fieldName, value) {
    this.setState(prevState => {
      const asset = { ...prevState.asset, [fieldName]: value }
      return {
        asset,
        errors: AssetForm.validate(asset),
      }
    })
  }

  handleChangeCurrent(value) {
    this.setState(prevState => ({
      current: value,
      asset: value
        ? _.omit(prevState.asset, 'purchaseDate', 'purchaseAmount')
        : _.omit(prevState.asset, 'currentValue'),
    }))
  }

  handleFocus(fieldName) {
    this.setState(() => ({ focusedInput: fieldName }))
  }

  handleSubmit(event) {
    event.preventDefault()
    this.setState(
      () => ({ errors: AssetForm.validate(this.state.asset) }),
      () => {
        const { errors, asset } = this.state
        if (
          typeof errors !== 'undefined' &&
          Validations.areErrorsPresent(errors)
        ) {
          const firstErrorFieldName = Validations.findFirstErrorFieldName(
            errors
          )
          firstErrorFieldName
            ? this.setState(() => ({
              allErrorsShown: true,
              focusedInput: firstErrorFieldName,
            }))
            : this.setState(() => ({ allErrorsShown: true }))
        } else {
          if (this.props.onSubmit) {
            this.props.onSubmit(asset)
          }
        }
      }
    )
    return false
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.asset) {
      this.setState(() => ({ asset: nextProps.asset }))
    }
  }

  render() {
    const { asset, current, allErrorsShown, focusedInput } = this.state
    const errors = this.state.errors === undefined ? {} : this.state.errors
    const idField =
      typeof asset.id === 'string'
        ? <HiddenField name='id' value={asset.id} />
        : null
    const today = moment().startOf('day')

    return (
      <form
        className='asset-form form form-aligned'
        onSubmit={this.handleSubmit}
      >
        <fieldset>
          <legend>General details</legend>

          {idField}
          <InputField
            name='name'
            type='text'
            label='Name'
            value={asset.name}
            errors={errors.name}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleChange('name', value)}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
        </fieldset>
        <fieldset>
          <InputField
            name='current'
            type='checkbox'
            label='I currently own this property'
            checked={current}
            onChange={(_, c) => this.handleChangeCurrent(c)}
          />
        </fieldset>
        {current
          ? <fieldset>
              <legend>Current value</legend>
              <CurrencyField
                name='currentValue'
                label='Current value'
                value={asset.currentValue}
                errors={errors.currentValue}
                forceErrorDisplay={allErrorsShown}
                onChange={value => this.handleChange('currentValue', value)}
                onFocus={this.handleFocus}
                focus={focusedInput}
              />
            </fieldset>
          : <fieldset>
              <legend>Future purchase</legend>
              <DateField
                name='purchaseDate'
                label='Purchase date'
                value={asset.purchaseDate}
                errors={errors.purchaseDate}
                forceErrorDisplay={allErrorsShown}
                minDate={today}
                onChange={value => this.handleChange('purchaseDate', value)}
                onFocus={this.handleFocus}
                focus={focusedInput}
              />
              <CurrencyField
                name='purchaseAmount'
                label='Purchase amount'
                value={asset.purchaseAmount}
                errors={errors.purchaseAmount}
                forceErrorDisplay={allErrorsShown}
                onChange={value => this.handleChange('purchaseAmount', value)}
                onFocus={this.handleFocus}
                focus={focusedInput}
              />
            </fieldset>}
        <fieldset>
          <legend>Future sale</legend>
          <DateField
            name='saleDate'
            label='Sale date'
            value={asset.saleDate}
            errors={errors.saleDate}
            forceErrorDisplay={allErrorsShown}
            minDate={today}
            onChange={value => this.handleChange('saleDate', value)}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
        </fieldset>
        <fieldset>
          <legend>Forecast capital growth</legend>
          <InputField
            name='forecastGrowth'
            type='number'
            min={0}
            label='% per year'
            value={asset.forecastGrowth}
            errors={errors.forecastGrowth}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleChange('forecastGrowth', value)}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
        </fieldset>
        <fieldset>
          <button type='submit' className='button button-primary'>
            Submit
          </button>
        </fieldset>
      </form>
    )
  }

  static validate(asset) {
    const name = []
      .concat(Validations.required(asset.name))
      .concat(Validations.minLength(asset.name, 3))
      .concat(Validations.maxLength(asset.name, 100))
    const forecastGrowth = Validations.required(asset.forecastGrowth)

    return { name, forecastGrowth }
  }
}

export default AssetForm
