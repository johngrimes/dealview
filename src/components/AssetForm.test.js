import React from 'react'
import { shallow } from 'enzyme'
import _ from 'lodash'
import MockDate from 'mockdate'

import AssetForm from './AssetForm.js'
import {
  validAsset1,
  validAsset2,
  validAsset3,
} from '../data/fixtures/asset.js'

beforeEach(() => MockDate.set('2017-01-01'))
afterEach(() => MockDate.reset())

describe('AssetForm', () => {
  for (const asset of [ validAsset1, validAsset2, validAsset3 ]) {
    it(`should submit the entered data for ${asset.name}`, () => {
      const props = { onSubmit: jest.fn() }
      const wrapper = shallow(<AssetForm {...props} />)
      if (
        typeof asset.purchaseDate === 'string' ||
        typeof asset.purchaseAmount === 'number'
      ) {
        wrapper.find({ name: 'current' }).prop('onChange')(null, false)
      }
      for (const field in asset) {
        const fieldWrapper = wrapper.find({ name: field })
        fieldWrapper.prop('onChange')(asset[field])
      }
      wrapper.find('form').prop('onSubmit')({ preventDefault: jest.fn() })
      expect(props.onSubmit).toHaveBeenCalledWith(asset)
    })
  }

  it('should be in current mode if a purchase date is present', () => {
    const props = { asset: validAsset3 }
    expect(typeof props.asset.purchaseDate).toBe('string')
    const wrapper = shallow(<AssetForm {...props} />)
    const fieldWrapper = wrapper.find({ name: 'purchaseDate' })
    expect(fieldWrapper.exists()).toBe(true)
  })

  it('should be in current mode if a purchase amount is present', () => {
    const props = { asset: validAsset3 }
    expect(typeof props.asset.purchaseAmount).toBe('string')
    const wrapper = shallow(<AssetForm {...props} />)
    const fieldWrapper = wrapper.find({ name: 'purchaseAmount' })
    expect(fieldWrapper.exists()).toBe(true)
  })

  it('should not be in current mode if purchase date and amount are absent', () => {
    const props = { asset: validAsset2 }
    expect(typeof props.asset.purchaseDate).toBe('undefined')
    expect(typeof props.asset.purchaseAmount).toBe('undefined')
    const wrapper = shallow(<AssetForm {...props} />)
    const fieldsExist = {
      purchaseDate: false,
      purchaseAmount: false,
      currentValue: true,
    }
    for (const field in fieldsExist) {
      const fieldWrapper = wrapper.find({ name: field })
      expect(fieldWrapper.exists()).toBe(fieldsExist[field])
    }
  })

  it('should not submit purchase details if current has been ticked', () => {
    const props = { asset: validAsset3, onSubmit: jest.fn() }
    expect(typeof props.asset.purchaseDate).toBe('string')
    expect(typeof props.asset.purchaseAmount).toBe('string')
    const wrapper = shallow(<AssetForm {...props} />)
    wrapper.find({ name: 'current' }).prop('onChange')(null, true)
    wrapper.find('form').prop('onSubmit')({ preventDefault: jest.fn() })
    expect(props.onSubmit).toHaveBeenCalledWith(
      _.omit(props.asset, 'purchaseDate', 'purchaseAmount')
    )
  })

  const requiredFields = { name: {}, forecastGrowth: { name: 'Some asset' } }
  for (const field in requiredFields) {
    it(`should require a value in the ${field} field`, () => {
      const props = {
        asset: requiredFields[field],
        onSubmit: jest.fn(),
      }
      const wrapper = shallow(<AssetForm {...props} />)
      wrapper.find('form').prop('onSubmit')({ preventDefault: jest.fn() })
      expect(props.onSubmit).not.toHaveBeenCalled()
      const fieldWrapper = wrapper.find({ name: field })
      expect(fieldWrapper.prop('errors').length).toBeGreaterThan(0)
      expect(fieldWrapper.prop('focus')).toBe(field)
      expect(fieldWrapper.prop('forceErrorDisplay')).toBe(true)
      expect(fieldWrapper).toMatchSnapshot()
    })
  }
})
