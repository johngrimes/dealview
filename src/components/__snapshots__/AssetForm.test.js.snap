// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetForm should require a value in the forecastGrowth field 1`] = `
<InputField
  errors={
    Array [
      "can't be blank",
    ]
  }
  focus="forecastGrowth"
  forceErrorDisplay={true}
  label="% per year"
  min={0}
  name="forecastGrowth"
  onChange={[Function]}
  onFocus={[Function]}
  type="number"
/>
`;

exports[`AssetForm should require a value in the name field 1`] = `
<InputField
  errors={
    Array [
      "can't be blank",
    ]
  }
  focus="name"
  forceErrorDisplay={true}
  label="Name"
  name="name"
  onChange={[Function]}
  onFocus={[Function]}
  type="text"
/>
`;
